src/plugins/embeddable/public/react_embeddable_system/react_embeddable_registry.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { i18n } from '@kbn/i18n';
+import React from 'react';
+import {
+  DefaultEmbeddableApi,
+  ReactEmbeddable,
+  ReactEmbeddableFactory,
+  ReactEmbeddableRegistration,
+} from './types';
+
+const registry: { [key: string]: ReactEmbeddableFactory<any, any> } = {};
+
+export const registerReactEmbeddableFactory = <
+  StateType extends unknown = unknown,
+  APIType extends DefaultEmbeddableApi = DefaultEmbeddableApi
+>(
+  key: string,
+  factory: ReactEmbeddableFactory<StateType, APIType>
+) => {
+  registry[key] = factory;
```

Comment:
This should throw if key already exists in register to prevent plugins from overwriting factories.

Reply:
Done in 1f94fd4e53c